### ISSUE TITLE
Extend tokenizer vocabulary with new words

### ISSUE THREAD

**Issue by anferico** [huggingface/tokenizers#627]
Suppose I have a pre-trained tokenizer, e.g. a `BertWordPieceTokenizer`, with its own vocabulary. My goal is to use it to tokenize some technical text which will likely contain unknown words (represented as "[UNK]" tokens).

Is there a way to fine-tune the tokenizer so that unknown words are automatically added to its vocabulary? I have found similar issues in the `transformers` repository ([transformers/issues/2691](https://github.com/huggingface/transformers/issues/2691) and [transformers/issues/1413](https://github.com/huggingface/transformers/issues/1413)), but what they suggest is to manually add unknown tokens, whereas I would like them to be added automatically.

Here's a pseudo-code representation of what I would need:
```
pre_trained_tokenizer = ...
vocab = pre_trained_tokenizer.get_vocab()

technical_text = [
  'some text with unknown words',
  'some other text with unknown words',
  ...
]

updated_tokenizer = pre_trained_tokenizer.train(
  technical_text,
  initial_vocabulary=vocab
)

new_vocab = updated_tokenizer.get_vocab()  # 'new_vocab' contains all words in 'vocab' plus some new words
```

Can I do that with `huggingface/tokenizers` and/or `huggingface/transformers`?
I thought it would be an easy thing to do, but I wasn't able to find anything useful.

**Comment by n1t0** [comment]
No that's not possible, you'll have to add the tokens manually indeed.

**Comment by anferico** [comment]
Thanks for the reply. Just to clarify, is it a missing feature of the library or is it a limitation of the tokenization algorithm?

**Comment by n1t0** [comment]
It depends on the specific tokenization algorithm, but the tokenizer doesn't save all the training state that would be needed to pick up the training back where it was initially left.

**Comment by jowagner** [comment]
Most off-the-shelf models have plenty of unused vocabulary entries that you could repurpose:
1. Train a new vocabulary on the target domain corpus from scratch
2. Find the new vocabulary entries that are not in the old vocabulary
3. If the number of new entries is outside the desired range change settings and/or corpus and repeat from step 1
4. Replace the last unused entries with the new entries
5. Fine-tune as usual

If your application needs some unused entries for itself you must of course leave a sufficient number of such entries.

**Comment by juanjucm** [comment]
> 
> 
> Here's a pseudo-code representation of what I would need:
> 
> ```
> pre_trained_tokenizer = ...
> vocab = pre_trained_tokenizer.get_vocab()
> 
> technical_text = [
>   'some text with unknown words',
>   'some other text with unknown words',
>   ...
> ]
> 
> updated_tokenizer = pre_trained_tokenizer.train(
>   technical_text,
>   initial_vocabulary=vocab
> )
> 
> new_vocab = updated_tokenizer.get_vocab()  # 'new_vocab' contains all words in 'vocab' plus some new words
> ```

Hi @anferico , I don't know if this is what you were looking for, but this could be a possible approach for your problem:

1. First, you need to extract tokens out of your data while applying the same preprocessing steps used by the tokenizer. To do so you can just use the tokenizer itself:
`new_tokens = tokenizer.basic_tokenizer.tokenize(' '.join(technical_text))`
2. Now you just add the new tokens to the tokenizer vocabulary:
`tokenizer.add_tokens(new_tokens)`
This method only adds new tokens, which means you don't have to worry about words already present in the tokenizer's vocab.

The result would be the tokenizer with your specific domain tokens along with the original tokenizer's vocabulary. Of course, you can just encapsulate this in a function and use it like you do in your pseudocode.

Remember that for your model to work, you will need to update the embedding layer with the new augmented vocabulary:
`model.resize_token_embeddings(len(tokenizer))`

Hope it helps!! :)

**Comment by dplaniel** [comment]
It does not seem like any of the PreTrainedTokenizer or PreTrainedTokenizerFast base classes actually inherit the SpecialTokensMixin that implements .add_tokens(), so this answer is something of a dead end.

**Comment by dplaniel** [comment]
I don't think .add_tokens() is implemented. https://github.com/huggingface/transformers/blob/cad61b68396a1a387287a8e2e2fef78a25b79383/src/transformers/tokenization_utils_base.py#L952

**Comment by Narsil** [comment]
> I don't think .add_tokens() is implemented. https://github.com/huggingface/transformers/blob/cad61b68396a1a387287a8e2e2fef78a25b79383/src/transformers/tokenization_utils_base.py#L952

You are pointing to a base class, so yes it's not implemented.

Real class: https://github.com/huggingface/transformers/blob/cad61b68396a1a387287a8e2e2fef78a25b79383/src/transformers/tokenization_utils_fast.py#L264

**Comment by harveyaot** [comment]
I think this tutorial shared by the official can help with your question [Training a new tokenizer from an old one](https://huggingface.co/learn/nlp-course/chapter6/2)

**Comment by savanth14** [comment]
@Narsil @n1t0 I tried to add new vocab to the existing mistral tokenizer vocab using the `add_tokens()` method. Everything went fine till I tried the extended vocab tokenizer for decoding the encoded text. I found that in the decoded text, the spaces are completely missing and all the decoded tokens are merged into a single string. Can you please help me resolve this issue. Here's the sample code:

import sentencepiece as spm

sp = spm.SentencePieceProcessor(model_file='mistral_tok.model')
tokenizer1 = transformers.AutoTokenizer.from_pretrained("mistralai/mistral-7b-v0.1")

vocab = [sp.id_to_piece(idx) for idx in range(sp.get_piece_size())]

new_tokens = set(vocab) - set(tokenizer1.vocab.keys())

tokenizer1.add_tokens(list(new_tokens))
# output: 14756

print("After adding new tokens, length of mistral tokenizer:", len(tokenizer1))
# output: 46756

tel_text = "నేను బాగున్నాను. మీరు ఏలా ఉన్నారు?" # original text

mistral_encode_ids = tokenizer1.encode(tel_text)

mistral_decode_text = tokenizer1.decode(mistral_encode_ids, skip_special_tokens=True)

print(mistral_decode_text)

# output: నేనుబాగున్నాను.మీరుఏలాఉన్నారు? # decoded text with missing spaces

**Comment by bezir** [comment]
@savanth14 Hello,

I want to state that when someone uses `add_tokens` method it just adds new token to the end of vocab. However, tokenizer is not all about the `vocab.json`. You should take into account the `merges.txt` file as well. The missing spaces are most likely due to some merge operations. When someone trains a `SentencePiece` tokenizer and merges vocabs, they may overlook to update merges.txt, which is wrong. On the other hand merging `merges.txt` files also leads to a suboptimal solution. I would advise 🤗's build-in `train_new_from_iterator` function however as far as I am concerned using `SentencePiece` leads to a solution where you can process more data to train tokenizer.

### Language Alingnment Tokenizer Possible Solutions:

- Use `train_new_from_iterator` and multilingual data to train a new tokenizer.
- Train a tokenizer using `SentencePiece` and language-to-be-aligned data, merge vocabs.json, merges.txt and other relevant files.
- This [solution](https://github.com/ymcui/Chinese-LLaMA-Alpaca/blob/main/scripts/merge_tokenizer/merge_tokenizers.py) works well. However, I cannot advise much more on this solution.

The most optimal solution might be the first one with enough data. However, the function consumes more ram in cpu than the other solutions. I can create a tokenizer by using more than 20 GB data and `SentencePiece`. However, that is not the case for the first solution.

One last thing, if the alphabet of the language-to-be-alingned is different than the base tokenizer's then merging `merges.txt` files may not be suboptimal.

**Comment by savanth14** [comment]
Hello @bezir, thanks for your comment. I figured out how to successfully merge two sentencepiece BPE tokenizers without losing the tokenization efficiency. Here's the code:
```
# Load the pre-trained tokenizer to be extended
original_tokenizer_path = hf_hub_download(repo_id="mistralai/mistral-7b-v0.1", filename="tokenizer.model", local_dir="original_tokenizer")
original_tokenizer_spm = sp_pb2_model.ModelProto()
original_tokenizer_spm.ParseFromString(open(original_tokenizer_path, "rb").read())

# Load the newly trained tokenizer
new_tokenizer_spm = sp_pb2_model.ModelProto()
new_tokenizer_spm.ParseFromString(open("/content/mistral_tel_tokenizer.model", "rb").read())


# Check if the new tokenizer contains english tokens
def contains_eng(text):
    eng_pattern = re.compile(r"[\u0020-\u007E]+")
    return True if eng_pattern.search(text) else False


original_tokenizer_tokenset = set(p.piece for p in original_tokenizer_spm.pieces)
print(f"Number of tokens before merge: {len(original_tokenizer_tokenset)}")
for p in new_tokenizer_spm.pieces:
    piece = p.piece
    if piece not in original_tokenizer_tokenset and not contains_eng(piece):
        new_p = sp_pb2_model.ModelProto().SentencePiece()
        new_p.piece = piece
        new_p.score = 0
        original_tokenizer_spm.pieces.append(new_p)
print(f"Number of tokens after merge: {len(original_tokenizer_spm.pieces)}")

# Save the extended tokenizer to a checkpoint
extended_tokenizer_save_path="/content/english-telugu-tokenizer"
os.makedirs(extended_tokenizer_save_path, exist_ok=True)
with open(os.path.join(extended_tokenizer_save_path, "tokenizer.model"), "wb") as f:
    f.write(original_tokenizer_spm.SerializeToString())
```

I adapted this code from this source: https://github.com/google/sentencepiece/blob/master/python/add_new_vocab.ipynb

However, I have a new problem now.  I trained a tiktoken's bytelevel tokenizer using the code from this repo: https://github.com/gautierdag/tokenizer-bench
When I merged this new tokenizer with a pre-trained one, the telugu encoding performance got degraded. The merged tokenizer is splitting telugu text into too many fine-grained tokens. Encoding-Decoding for english, and decoding for telugu are working fine. Here's the code I used for merging: 
```
import json

def merge_tokenizers(file1, file2, output_file):
    # Load the tokenizers
    with open(file1, 'r') as f:
        tokenizer1 = json.load(f)
    with open(file2, 'r') as f:
        tokenizer2 = json.load(f)

    # Get the maximum rank in tokenizer1's vocab
    max_rank = max(tokenizer1['model']['vocab'].values())

    # Combine the vocabs and merges
    combined_vocab = tokenizer1['model']['vocab'].copy()
    for token, rank in tokenizer2['model']['vocab'].items():
        if token not in combined_vocab:
            combined_vocab[token] = len(combined_vocab) + 1


    combined_merges = tokenizer1['model']['merges'].copy()
    for merge in tokenizer2['model']['merges']:
        if merge not in combined_merges:
            combined_merges.append(merge)

    # combined_merges = tokenizer1['model']['merges'].copy()
    # combined_merges.extend(merge for merge in tokenizer2['model']['merges'] if merge not in combined_merges)

    # Update the vocab and merges in tokenizer1
    tokenizer1['model']['vocab'] = combined_vocab
    tokenizer1['model']['merges'] = combined_merges

    # Save the updated tokenizer
    with open(output_file, 'w') as f:
        json.dump(tokenizer1, f)

# Usage
merge_tokenizers("/content/gpt_32k.json", "/content/telugu_tokenizer_tiktoken.json", 'tokenizer_18.json')
```
The new tokenizer's training corpus is completely different from that of the pre-trained one. Also, I made sure to remove any duplicate merges. Still the encoding performance is poor.

**Comment by ArthurZucker** [comment]
Few things here. The extra spaces that are removed is because you are using the `legacy` behaviour which was fixed in `main`. 
Init the tokenizer with `legacy=False`.  
Missing spaces can also be fixed by setting `normalized=False` when adding the token as an `AddedToken`: 
```python 
from transformers import AddedToken, AutoTokenizer
tokenizer = transformers.AutoTokenizer.from_pretrained("mistralai/mistral-7b-v0.1")
tokenizer.add_tokens(AddedToken("<bbb>",normalized=True), True)
tokenizer.decode(tokenizer.encode(". <bbb>"))
'<s> .<bbb>'
```
vs
```python
from transformers import AddedToken, AutoTokenizer
tokenizer = transformers.AutoTokenizer.from_pretrained("mistralai/mistral-7b-v0.1")
tokenizer.add_tokens(AddedToken("<bbb>",normalized=False), True)
tokenizer.decode(tokenizer.encode(". <bbb>"))
'<s> . <bbb>'
```

**Comment by ArthurZucker** [comment]
> The merged tokenizer is splitting telugu text into too many fine-grained tokens 

try setting the `ignore_merges` parameter in `BPETokenizer` to `True`. See [this PR: ](https://github.com/huggingface/tokenizers/pull/1493)
